### PR TITLE
Make sure that ResonanceMolSupplier substructure matches are uniquified consistently

### PR DESCRIPTION
**Pull request by ptosco**
Today I have built the RDKit with GCC 9.3.0 on a machine with 40 CPU cores and I had a failure in `testCrambin` (`Code/GraphMol/Wrap/rough_test.py`). Upon investigation this turned out to be caused by the algorithm that only keeps unique substructure matches not being deterministic, and hence liable to returning different matches depending on the substructure match concurrency.
This PR implements a robust solution, which I believe is also more elegant and computationally efficient compared to the one that was in place before.
Now match uniquification always returns the match vector with indices in the minimum sorted order across all the possible permutations. I have adjusted `testCrambin` results accordingly.
I am surprised that the test has not failed before.

### CODE REVIEW

**Comment by ptosco** [comment]
The Ubuntu 16.04 build is failing with a strange linker error in `libxml2`, but I don't think that's related to this PR.

File: Code/GraphMol/Substruct/SubstructMatch.cpp [hunk 1]
```diff
@@ -130,16 +130,59 @@ void MatchSubqueries(const ROMol &mol, QueryAtom::QUERYATOM_QUERY *q,
...
+  }
+  if (shouldInsert) {
+    matches.insert(m);
+  }
+  return shouldInsert;
+}
+
+template <
+    typename Container,
+    typename std::enable_if<
```
Comment (greglandrum) [inline Code/GraphMol/Substruct/SubstructMatch.cpp:166]: Is this complication necessary?
It's cute to use language features, but isn't there a simpler way to express this?
(Note, there may not be... I haven't banged my head against the syntax and the reference material in order to figure out what this does yet)

File: Code/GraphMol/Substruct/SubstructMatch.cpp [hunk 2]
```diff
@@ -130,16 +130,59 @@ void MatchSubqueries(const ROMol &mol, QueryAtom::QUERYATOM_QUERY *q,
...
+        std::is_same<MatchVectType, typename Container::value_type>::value,
+        int>::type = 0>
+void mergeMatchVect(std::set<MatchVectType> &matches,
+                    const Container &matchesTmp,
+                    const ResSubstructMatchHelperArgs_ &args) {
+  for (auto it = matchesTmp.begin();
```
Comment (greglandrum) [inline Code/GraphMol/Substruct/SubstructMatch.cpp:172]: I'm not a fan of this syntax. I think it's clearer to just iterate over the elements of `matchesTmp` using the range iteration syntax and then have a break in the loop for when `matches.size()>=args.params.maxMatches`

**Review by greglandrum — CHANGES_REQUESTED** [review]
a couple small changes suggested.
Thanks for tracking this down and fixing it.

**Comment by ptosco** [comment]
@greglandrum I think I have addressed all of your comments; please let me know if this is not the case. Thanks for reviewing!

**Review by greglandrum — APPROVED** [review]

**Comment by greglandrum** [comment]
An aside: the force-push seems to have really screwed up github's ability to show the diffs in a useable form.

**Pull request merged**